File: 2024/python/day17.py

from dataclasses import dataclass
import itertools
from data.day17 import PUZZLE_INPUT, TEST_INPUT

@dataclass
class Instruction:
	opcode: int
	operand: int

@dataclass
class Program:
    registers:   dict[str,int]
    instructions: list[Instruction]
    insPointer:   int
    halted:       bool
    output:       list[int]

    def __init__(self, input):

        self.insPointer = 0
        self.halted = False
        self.output = []

        frags = input.split("\n\n")
        self.registers = {}

        for register in frags[0].splitlines():
            _, name, val = register.split()
            self.registers[name[:-1]] = int(val)


        ns = frags[1][9:].split(",")

        self.instructions = []

        for (opcode, operand) in itertools.batched(ns, 2):
            self.instructions.append(Instruction(int(opcode), int(operand)))

    def run(self):
        while not self.halted:
            self.execute()
            
    def has_next_instruction(self):
        self.insPointer += 1
        if int(self.insPointer) == len(self.instructions):
            self.halted = True
            return False
        return True

    def execute(self):

        ins = self.instructions[self.insPointer]

        match ins.opcode:
            case 0: #adv
                nr = self.registers["A"]
                dr = 2 ** self.combo_operand(ins.operand)
                self.registers["A"] = nr // dr

            case 1: #bxl
                self.registers["B"] ^= ins.operand

            case 2: #bst
                self.registers["B"] = self.combo_operand(ins.operand) % 8
               
            case 3: #jnx
                if self.registers["A"] != 0:
                    self.insPointer = ins.operand
                

            case 4: #bxc
                self.registers["B"] ^= self.registers["C"]

            case 5: #out
                # print("5", self.combo_operand(ins.operand), self.combo_operand(ins.operand)%8)
                self.output.append(self.combo_operand(ins.operand)%8)

            case 6: #bdv
                nr = self.registers["A"]
                dr = 2 ** self.combo_operand(ins.operand)
                self.registers["B"] = nr // dr

            case 7: #cdv
                nr = self.registers["A"]
                dr = 2 ** self.combo_operand(ins.operand)
                self.registers["C"] = nr // dr

        if ins.opcode != 3 or self.registers["A"] == 0:
            return self.has_next_instruction()
        
        return False

    def combo_operand(self, operand: int):

        if operand >= 0 and operand <= 3:
            return operand
        else:
            match operand:
                case 4:
                    return self.registers["A"]
                case 5:
                    return self.registers["B"]
                case 6:
                    return self.registers["C"]
                case 7:
                    return -1
                case _:
                    return -1
            
    def reset(self, reg_a_value: int):
        self.registers["A"] = int(reg_a_value)
        self.registers["B"] = 0
        self.registers["C"] = 0
        self.insPointer = 0
        self.halted = False
        self.output = []


    def compute_value_for_register_a(self):

        x = 7
        le = 0
        a = 7

        while True:

            for a1 in range(x * 8, x * 9):

                self.reset(a1)
                self.run()

                le = len(self.output)

                if self.output[:le] == expected[-le:]:
                    x = a * 8
                    a = a1
                    break
    
            if le == 16:
                self.registers["A"] = a
                return


# sequence
# 2 4 - b = a % 8
# 1 3 - b = b ^ 3
# 7 5 - c = a // (2 ** b)
# 0 3 - a = a ^ 8
# 1 4 - b = b ^ 4
# 4 7 - b = b ^ c (b is 2)
# 5 5 - write b mod 8 (this is 2)
# 3 0 - reset instruction pointer

expected = [2, 4, 1, 3, 7, 5, 0, 3, 1, 4, 4, 7, 5, 5, 3, 0]



def solve(part2, input):

    program = Program(input)
    
    if part2:
        program.compute_value_for_register_a()

    program.run()

    if part2:
        print(program.registers["A"])
    else:
        print(",".join(map(str, program.output)))

solve(False, TEST_INPUT)
solve(True, PUZZLE_INPUT)

File: 2024/go/17/day17.go
package Day17

import (
	"fmt"
	"math"
	"os"
	"slices"
	"strconv"
	"strings"
)

type Instruction struct {
	opcode  int64
	operand int64
}

type Program struct {
	registers    map[string]int64
	instructions []Instruction
	insPointer   int64
	halted       bool
	output       []string
}

func (prog *Program) Run() {
	for !prog.halted {
		prog.Execute()
	}
}

func (prog *Program) GetInstruction() Instruction {
	return prog.instructions[prog.insPointer]
}

func (prog *Program) NextInstruction() {
	prog.insPointer += 1
	if int(prog.insPointer) == len(prog.instructions) {
		// fmt.Println(strings.Join(prog.output, ","))
		prog.halted = true
	}
}

func (prog *Program) Execute() {

	ins := prog.GetInstruction()

	switch ins.opcode {
	case 0: //adv
		nr := prog.registers["A"]
		dr := math.Pow(2, float64((prog.comboOperand(ins.operand))))
		prog.registers["A"] = nr / int64(dr)

	case 1: //bxl
		prog.registers["B"] ^= ins.operand

	case 2: //bst
		prog.registers["B"] = prog.comboOperand(ins.operand) % 8

	case 3: //jnx
		if prog.registers["A"] != 0 {
			prog.insPointer = ins.operand
		}

	case 4: //bxc
		prog.registers["B"] ^= prog.registers["C"]

	case 5: //out
		fmt.Println("5", prog.comboOperand(ins.operand), prog.comboOperand(ins.operand)%8)
		prog.output = append(prog.output, strconv.Itoa(int(prog.comboOperand(ins.operand)%8)))
		// fmt.Println(strings.Join(prog.output, ","))

	case 6: //bdv
		nr := prog.registers["A"]
		dr := math.Pow(2, float64((prog.comboOperand(ins.operand))))
		prog.registers["B"] = nr / int64(dr)

		fmt.Println("bdv", prog.registers["A"], (prog.comboOperand(ins.operand)), dr, prog.registers["B"])

	case 7: //cdv
		nr := prog.registers["A"]
		dr := math.Pow(2, float64((prog.comboOperand(ins.operand))))
		prog.registers["C"] = nr / int64(dr)

	}

	if ins.opcode != 3 || prog.registers["A"] == 0 {
		prog.NextInstruction()
	}

}

func (prog *Program) comboOperand(operand int64) int64 {

	if operand >= 0 && operand <= 3 {
		return operand
	} else {
		switch operand {
		case 4:
			return prog.registers["A"]
		case 5:
			return prog.registers["B"]
		case 6:
			return prog.registers["C"]
		case 7:
			fmt.Println("Reserved!")
			return -1
		default:
			return -1
		}
	}
}

func (program *Program) reset(reg_a_value int64) {
	program.registers["A"] = int64(reg_a_value)
	program.registers["B"] = 0
	program.registers["C"] = 0
	program.insPointer = 0
	program.halted = false
	program.output = []string{}
}

func parseInput(file_path string) Program {
	bs, _ := os.ReadFile(file_path)
	input := string(bs)

	frags := strings.Split(input, "\n\n")
	registers_str := strings.Split(frags[0], "\n")
	registers := map[string]int64{}

	for _, register := range registers_str {
		fs := strings.FieldsFunc(register, func(r rune) bool {
			return r == ':' || r == ' '
		})

		val, _ := strconv.ParseInt(fs[2], 10, 64)
		registers[fs[1]] = val
	}

	ns := strings.Split(frags[1][9:], ",")

	instructions := []Instruction{}

	for pair := range slices.Chunk(ns, 2) {
		opcode, _ := strconv.ParseInt(pair[0], 10, 64)
		operand, _ := strconv.ParseInt(pair[1], 10, 64)

		instructions = append(instructions, Instruction{opcode, operand})
	}

	return Program{registers, instructions, 0, false, []string{}}

}

func Part1() {

	file_path := "./17/test_input1"
	program := parseInput(file_path)

	expected := "2,4,1,3,7,5,0,3,1,4,4,7,5,5,3,0"

	// nums := []int64{2, 4, 1, 3, 7, 5, 0, 3, 1, 4, 4, 7, 5, 5, 3, 0}

	expected_total := int64(2413750314475530)

	nums := []int64{5, 4, 6, 5, 2, 7, 4, 7, 6, 0, 3, 4, 5, 4, 7, 0}

	total := int64(0)

	for x := int64(777777); x > 0; x-- {

		total = 0

		e := int64(1)

		// for y := 0; y < 16; y++ {
		// 	nums[15-y] = x / e % 10
		// 	e *= 10
		// }

		for y := 3; y <= 8; y++ {
			nums[y] = x / e % 10
			e *= 10
		}

		for p, num := range nums {
			total += num * int64(math.Pow(8, float64(p+1)))

			if total > expected_total {
				break
			}

			s := strconv.Itoa(int(total))

			if strings.HasPrefix(s, "241375031") {
				fmt.Println(s, nums)
			}

		}

		if total == int64(expected_total) {
			print("Here you go!")
		}

	}

	// expected := "0,3,5,4,3,0"

	program.reset(total)
	program.Run()

	fmt.Println(program.output)

	if strings.Join(program.output, ",") == expected {
		print("Done!")

	}

}
